execution-engine/src/fs.rs

```diff
@@ -287,6 +335,44 @@ impl InodeImpl {
         }
     }
 
+    /// Check if it is a service
+    #[inline]
+    pub(crate) fn is_service(&self) -> bool {
+        match self {
+            Self::NativeModule(..) => true,
+            _ => false,
+        }
+    }
+
+    /// Try to parse the input of a service. Return true if it is a valid input.
+    #[inline]
+    pub(crate) fn service_valid_input(&self) -> FileSystemResult<bool> {
```

Comment:
When calling this function, we fully parse the input. Do we save the output of that parsing, so we won't have to parse it again when we consume it?

Reply:
The native module might need to parse it again, depending on how the native module implements the interface `try_parse`, where the native module could buffer previous results in their own space.

I am using this function as a `try-parse` pattern. When it is successful (returning true), it means there is enough input for the native module. This is because we are triggering the native module on writing to a special file, and there is no guarantee a single write (in high-level source code) will be done in one invocation of `fd_write` (in the wasi level). Essentially, this function is called on every invocation to check if there is enough.

Reply:
Understood.
So the native module might save the result of a successful `try-parse`. That's the preferred way to avoid multiple successful parses.